### ISSUE TITLE
Clarification on KL-Divergence Term

### ISSUE THREAD

**Issue by ShayekhBinIslam** [open-thought/tiny-grpo#5]
The `action_mask` seems to be applied twice for the `kl` term. 

First:
https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L19

Second:
https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L64

If you kindly clarify, it would be helpful to understand better as without the first step, the results become different.

If the calculation is correct, the effective KL term should be `masked_mean(kl, action_mask, dim=-1).mean()` rather than `kl.mean()` here: https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L65

**Comment by ktiwary2** [comment]
Has this been fixed in the main code? What was the final change here?

**Comment by oodrand** [comment]
> Has this been fixed in the main code? What was the final change here?

Commit history does not reflect any fixes being done, you'll have to do the changes locally

**Comment by billhhh** [comment]
Hi

> The `action_mask` seems to be applied twice for the `kl` term.
> 
> First:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L19)
> 
> Line 19 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  log_ratio = log_ratio * action_mask 
> Second:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L64)
> 
> Line 64 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  loss = masked_mean(loss, action_mask, dim=-1).mean() 
> If you kindly clarify, it would be helpful to understand better as without the first step, the results become different.
> 
> If the calculation is correct, the effective KL term should be `masked_mean(kl, action_mask, dim=-1).mean()` rather than `kl.mean()` here:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L65)
> 
> Line 65 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  return loss, kl.mean()

Hello, I do not think it is a bug as kl in the loss has not been used anyway, so it is not duplicated.
Correct me if I am wrong.

**Comment by billhhh** [comment]
> The `action_mask` seems to be applied twice for the `kl` term.
> 
> First:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L19)
> 
> Line 19 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  log_ratio = log_ratio * action_mask 
> Second:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L64)
> 
> Line 64 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  loss = masked_mean(loss, action_mask, dim=-1).mean() 
> If you kindly clarify, it would be helpful to understand better as without the first step, the results become different.
> 
> If the calculation is correct, the effective KL term should be `masked_mean(kl, action_mask, dim=-1).mean()` rather than `kl.mean()` here:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L65)
> 
> Line 65 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  return loss, kl.mean()

Hi, the kl.mean() has not been used, right?

**Comment by billhhh** [comment]
> The `action_mask` seems to be applied twice for the `kl` term.
> 
> First:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L19)
> 
> Line 19 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  log_ratio = log_ratio * action_mask 
> Second:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L64)
> 
> Line 64 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  loss = masked_mean(loss, action_mask, dim=-1).mean() 
> If you kindly clarify, it would be helpful to understand better as without the first step, the results become different.
> 
> If the calculation is correct, the effective KL term should be `masked_mean(kl, action_mask, dim=-1).mean()` rather than `kl.mean()` here:
> 
> [tiny-grpo/loss.py](https://github.com/open-thought/tiny-grpo/blob/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c/loss.py#L65)
> 
> Line 65 in [eafedd7](/open-thought/tiny-grpo/commit/eafedd78ff86dbb724a3dd21bb04ab6523ac8f3c)
> 
>  return loss, kl.mean()

I do not think it is a bug as mask is to calc the kl and also for the loss calculation. There is no conflict as the mask will take out the useless positions.
Correct me if I am wrong.
